File: ruby/teller.rb
#!/usr/bin/env ruby
require "sinatra"
require "net/http"
require "base64"
require "openssl"
require "uri"

set :bind, "0.0.0.0"
set :port, (ENV["PORT"] || 8001).to_i

APP_ID      = ENV["APP_ID"]
ENVIRONMENT = ENV["ENV"] || "sandbox"
CERT        = ENV["CERT"]
CERT_KEY    = ENV["CERT_KEY"]

abort "APP_ID must be set (e.g. APP_ID=app_xxx)" if APP_ID.nil? || APP_ID.strip.empty?
if %w[development production].include?(ENVIRONMENT) && (CERT.nil? || CERT_KEY.nil?)
  abort "CERT and CERT_KEY must be set when ENV=#{ENVIRONMENT}"
end

BASE_TELLER_URL = "https://api.teller.io"

def static_dir
  File.expand_path("../static", __dir__)
end

def proxy_to_teller(path)
  uri = URI.join(BASE_TELLER_URL + "/", path)

  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = true
  if CERT && CERT_KEY
    http.cert = OpenSSL::X509::Certificate.new(File.read(CERT))
    http.key  = OpenSSL::PKey.read(File.read(CERT_KEY)) # don’t assume RSA
  end

  method = request.request_method.capitalize
  klass = Net::HTTP.const_defined?(method) ? Net::HTTP.const_get(method) : Net::HTTP::Get
  upstream_req = klass.new(uri.request_uri)

  # Preserve Content-Type
  upstream_req["Content-Type"] = request.content_type if request.content_type
  upstream_req["Accept"] = "application/json"

  # Forward body for POST/PUT/PATCH
  if %w[POST PUT PATCH].include?(request.request_method)
    body = request.body.read
    upstream_req.body = body unless body.nil? || body.empty?
  end

  # Rewrite Authorization: <token> -> Basic base64(token:)
  if (raw = request.env["HTTP_AUTHORIZATION"])
    token = raw.strip
    upstream_req.basic_auth(token, "")
  end

  upstream_resp = http.request(upstream_req)

  status upstream_resp.code.to_i
  response.headers.clear
  upstream_resp.each_header do |k, v|
    next if %w[transfer-encoding connection content-encoding content-length].include?(k.downcase)
    response.headers[k] = v
  end
  content_type upstream_resp["content-type"] if upstream_resp["content-type"]

  upstream_resp.body
end

# Proxy verbs
%w[get post put delete patch options].each do |verb|
  send(verb, "/api/*") do
    subpath = params["splat"].first
    proxy_to_teller(subpath)
  end
end

# Root with substitutions
get "/" do
  html = File.read(File.join(static_dir, "index.html"))
  html = html.gsub("{{ app_id }}", APP_ID)
  html = html.gsub("{{ environment }}", ENVIRONMENT)
  content_type "text/html"
  html
end

# Static assets
get "/static/*" do
  file = params["splat"].first
  send_file File.join(static_dir, file)
end

File: php/teller.php
<?php
// Run: APP_ID=app_xxx php -S localhost:8001 teller.php

$appId   = getenv("APP_ID") or die("APP_ID must be set\n");
$env     = getenv("ENV") ?: "sandbox";
$cert    = getenv("CERT");
$certKey = getenv("CERT_KEY");

if (in_array($env, ["development", "production"]) && (!$cert || !$certKey)) {
    die("CERT and CERT_KEY must be set when ENV=$env\n");
}

$baseDir = __DIR__ . "/../static";
$uri     = parse_url($_SERVER["REQUEST_URI"], PHP_URL_PATH);
$method  = $_SERVER["REQUEST_METHOD"];

// ---------- Root ----------
if ($uri === "/") {
    $html = str_replace(
        ["{{ app_id }}", "{{ environment }}"],
        [$appId, $env],
        file_get_contents("$baseDir/index.html")
    );
    header("Content-Type: text/html");
    echo $html;
    exit;
}

// ---------- Static ----------
if (strpos($uri, "/static/") === 0) {
    $path = substr($uri, 8); // after "/static/"
    $file = realpath("$baseDir/$path");
    if ($file && file_exists($file)) {
        header("Content-Type: " . mime_content_type($file));
        readfile($file);
    } else {
        http_response_code(404);
        echo "Not found";
    }
    exit;
}

// ---------- Proxy /api/* ----------
if (strpos($uri, "/api/") === 0) {
    $url = "https://api.teller.io/" . substr($uri, 5); // after "/api/"
    $headers = [];

    if (isset($_SERVER["HTTP_AUTHORIZATION"])) {
        $token = trim($_SERVER["HTTP_AUTHORIZATION"]);
        $headers[] = "Authorization: Basic " . base64_encode("$token:");
    }

    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_CUSTOMREQUEST => $method,
        CURLOPT_HTTPHEADER    => $headers,
        CURLOPT_RETURNTRANSFER=> true,
        CURLOPT_POSTFIELDS    => file_get_contents("php://input"),
    ]);
    if ($cert && $certKey) {
        curl_setopt($ch, CURLOPT_SSLCERT, $cert);
        curl_setopt($ch, CURLOPT_SSLKEY, $certKey);
    }

    $respBody = curl_exec($ch);
    $status   = curl_getinfo($ch, CURLINFO_RESPONSE_CODE);
    curl_close($ch);

    header("Content-Type: application/json");
    http_response_code($status);
    echo $respBody;
    exit;
}

// ---------- Default ----------
http_response_code(404);
echo "Not found";